llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp

```diff
@@ -382,6 +416,74 @@ PreservedAnalyses CompileTimePropertiesPass::run(Module &M,
                                   : PreservedAnalyses::all();
 }
 
+void CompileTimePropertiesPass::parseAlignmentAndApply(
+    Module &M, IntrinsicInst *IntrInst) {
+  LLVMContext &Ctx = M.getContext();
+  unsigned MDKindID = Ctx.getMDKindID(SPIRV_DECOR_MD_KIND);
+
+  // Get the global variable with the annotation string.
+  const GlobalVariable *AnnotStrArgGV = nullptr;
+  const Value *IntrAnnotStringArg = IntrInst->getArgOperand(1);
+  if (auto *GEP = dyn_cast<GEPOperator>(IntrAnnotStringArg))
+    if (auto *C = dyn_cast<Constant>(GEP->getOperand(0)))
+      AnnotStrArgGV = dyn_cast<GlobalVariable>(C);
+  if (!AnnotStrArgGV)
+    return;
+
+  std::optional<StringRef> AnnotStr = getGlobalVariableString(AnnotStrArgGV);
+  if (!AnnotStr)
+    return;
+
+  // parse properties string to decoration-value pairs
+  auto properties = parseSYCLPropertiesString(M, IntrInst);
+
+  SmallVector<Value *, 8> userList;
+  SmallVector<Instruction *, 4> instList;
+  // check if used by a load or store instructions
+  for (auto val : IntrInst->users()) {
+    // if castInst, push sussessors
+    if (auto cast = dyn_cast<CastInst>(val)) {
+      for (auto sussessor : cast->users())
+        userList.push_back(sussessor);
```

Comment:
nit: We can check if the inst is a load/store and then push only if it is? Also, is there a possibility of having a chain of casts?

Reply:
There was a possibility of having a chain of cast since there could be many addrspacecast and pointer cast in between. After we moved to opaque pointers these situations should be mitigated